File: java/1329_SortTheMatrixDiagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        for (int i = 0; i < mat.length; ++i)
            storeAndSortDiagonal(mat, i, 0);
        
        for (int j = 0; j < mat[0].length; ++j)
            storeAndSortDiagonal(mat, 0, j);
        
        return mat;
    }
    
    public void storeAndSortDiagonal(int[][] mat, int row, int col) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>((a, b) -> a - b);
        
        int i = row, j = col;
        while (i < mat.length && j < mat[0].length) 
            pq.add(mat[i++][j++]);
        
        while (!pq.isEmpty()) 
            mat[row++][col++] = pq.poll();
    }
}

File: swift/1329_SortTheMatrixDiagonally.swift
class Solution {
    func diagonalSort(_ mat: [[Int]]) -> [[Int]] {
        func storeAndSortDiagonal(_ row: Int, _ col: Int) {
            var i = row, j = col
            var diagonal = [Int]()
            while i < N && j < M {
                diagonal.append(A[i][j])
                i += 1
                j += 1
            }
            
            diagonal = diagonal.sorted() { $0 < $1 }
            i = row
            j = col
            for n in diagonal {
                A[i][j] = n
                i += 1
                j += 1
            }
        }
        
        var A = mat
        var N = A.count, M = A[0].count
        for i in 0..<N {
            storeAndSortDiagonal(i, 0)
        }
        
        for j in 0..<M {
            storeAndSortDiagonal(0, j)
        }
        
        return A
    }
}
